docs/source/sparse.rst

```diff
@@ -1,146 +1,382 @@
-
-You can also construct hybrid sparse tensors, where only the first n
-dimensions are sparse, and the rest of the dimensions are dense.
-
-    >>> i = torch.LongTensor([[2, 4]])
-    >>> v = torch.FloatTensor([[1, 3], [5, 7]])
-    >>> torch.sparse.FloatTensor(i, v).to_dense()
-     0  0
-     0  0
-     1  3
-     0  0
-     5  7
-    [torch.FloatTensor of size 5x2]
+    >>> i = [[0, 2], [1, 0], [1, 2]]
+    >>> v =  [3,      4,      5    ]
+    >>> s = torch.sparse_coo_tensor(list(zip(*i)), v, (2, 3))
+    >>> # Or another equivalent formulation to get s
+    >>> s = torch.sparse_coo_tensor(torch.tensor(i).t(), v, (2, 3))
+    >>> torch.sparse_coo_tensor(i.t(), v, torch.Size([2,3])).to_dense()
+    tensor([[0, 0, 3],
+            [4, 0, 5]])
+
+Sparse COO tensors support indexing operations (except slicing in
+sparse dimensions):
+
+    >>> s[1]
+    tensor(indices=tensor([[0, 2]]),
+           values=tensor(  [4, 5]),
+           size=(3,), nnz=2, layout=torch.sparse_coo)
+    >>> s[1, 0]
+    tensor(4)
 
 An empty sparse tensor can be constructed by specifying its size:
 
-    >>> torch.sparse.FloatTensor(2, 3)
-    SparseFloatTensor of size 2x3 with indices:
-    [torch.LongTensor with no dimension]
-    and values:
-    [torch.FloatTensor with no dimension]
+    >>> torch.sparse_coo_tensor(size=(2, 3))
+    tensor(indices=tensor([], size=(2, 0)),
+           values=tensor([], size=(0,)),
+           size=(2, 3), nnz=0, layout=torch.sparse_coo)
+
+.. _sparse-hybrid-coo-docs:
+
+Hybrid COO tensors
+------------------
+
+You can also construct hybrid tensors, where only the first dimensions
+are sparse, and the rest of the dimensions are dense. For instance,
+
+    >>> i = [[2, 4]]
+    >>> v = [[1, 3], [5, 7]]
+    >>> s = torch.sparse_coo_tensor(i, v, (5, 2))
+    >>> s.to_dense()
+    tensor([[0, 0],
+            [0, 0],
+            [1, 3],
+            [0, 0],
+            [5, 7]])
+
+where ``s`` is a (1+1)-dimensional :ref:`hybrid tensor
+<sparse-hybrid-docs>`:
+
+    >>> s[0]
+    tensor([0, 0])
+    >>> s[2]
+    tensor([1, 3])
+
+Notice that slicing in dense dimensions is supported (except with
+negative step):
+
+    >>> s[2, 1:]
+    tensor([3])
+
+In general, a (M + K)-dimensional coalesced sparse COO tensor ``s``
+with the number of specified elements ``nse`` has the following
+invariants:
+
+    >>> s = s.coalesce()
+    >>> M, K = s.sparse_dim(), s.dense_dim()
+    >>> assert M + K == len(s.shape) == s.ndim
+    >>> assert s.indices().shape == (M, nse)
+    >>> assert s.values().shape == (nse,) + s.shape[M:]
+
+Since ``s.indices()`` is always a 2D tensor, the smallest sparse
+dimension M = 1.  Therefore, representation of a sparse tensor of M =
+0 is simply a dense tensor.
+
+.. _sparse-uncoalesed-coo-docs:
+
+Uncoalesced sparse COO tensors
+------------------------------
+
+Torch sparse COO tensor format permits *uncoalesced* sparse tensors,
+where there may be duplicate coordinates in the indices; in this case,
```

Comment:
We should be careful about our language regarding uncoalesced tensors. I think in another place in these documents they're described as definitely having duplicative indices. Here a tensor is uncoalesced if it MAY have duplicative entries. 

Isn't it uncoalesced if it DOES have duplicative entries but is_coalesced only returns True if it's known to have NO duplicative entries? There may be some impedance with that function returning False when it means "maybe".

Reply:
When PyTorch constructs a sparse COO tensor, it will have `coalesced` bit set to False, even when indices do not contain duplicated indices and are ordered lexicographically. Only the `coalesce()` method can return a sparse tensor with the `coalesced` bit set to True.

So, `is_coalesced()` can return False even when the sparse tensor data is actually coalesced. When `is_coalesced()` returns True, then it is guaranteed that the sparse tensor data is coalesced.

In conclusion,  "a tensor is uncoalesced if it MAY have duplicative entries or the entries MAY not be ordered" is correct.

Reply:
I get that... I guess I'm asking if is_coalesced is an implementation-specific term only used in PyTorch's implementation of the coordinate list layout? Or is it a term of art? 

Also, if it is our own term, we may want to pick a different name in the future since "coalesced" and "uncoalesced" data access patterns already exist.

Reply:
"coalescing" may have many meanings: merging blocks, finding the first non-null value, etc.
So, I think, here the "coalesced tensor" is pytorch implementation specific term. I don't know what is PyTorch history related to the "uncoalesced tensor" term.